test-network-function/test-network-function
# Initial Ginkgo Generic and CNF-Specific Test Cases

Abstracts a set of test cases:
1) generic_cnf_tests:
   * ping 10x from test partner to pod under test on the default network
   * ping 10x from pod under test to test partner on the default network
   * ping 10x from test partner to pod under test on the Multus overlay network
2) cisco_kiknos_cnf_specific_tests:
   * create a tunnel and make sure a new tunnel shows up as "Connected" in the
     itc console
   * ping across the IPSEC tunnel from the test partner to the pod under test
   * send UDP packets across the IPSEC tunnel from the test partner to the
     pod under test

This implementation takes a dual driver approach:
1) generic_cnf_tests:  The existing tnf library was sufficient to appropriately
   run the tests, so it was used.  However, since back propagation of chained
   results was not yet merged, one-shot approaches were preferred.  Thus, the
   following were abstracted:
     * ocipaddr:  Utilizes "oc exec" and "ip addr" commands to get interface
     info for the default network.
     * ocping:  Utilizes "oc exec" and "ping" to ping from one pod to another.
   When the new chaining API(s) are available, these abstractions will
   disappear.
2) cisco_kiknos_cnf_specific_tests: The existing tnf library was found
   insufficient for matching EOF.  Unfortunately, the Cisco "itc" CLI tool
   largely relies on commands without output ("itc create", "itc ping", and
   "itc senddata") and commands that have no clear end ("itc ike" and
   "itc summary").  Thus, a temporary "itc" client library was abstracted which
   could adequately handle interaction with the "itc" CLI client.
   Additionally, to separate responsibilities, an "oc" client was created, as
   well as a lightweight "subprocess" client.  These were created to prevent
   code duplication across the client calls.  This library is temporary, and
   will be removed once "tnf" changes to allow matching on EOF are merged.

Test Configuration is added in "test-network-function/test.conf".  It is
described in detail in "test-network-function/configuration.go".  Several
Makefile targets were added to support building and running the code.  Namely,
"make cnftests" was added, which builds all go code, builds the tests, then
runs the tests.

Signed-off-by: Ryan Goulding <rgouldin@redhat.com>
